### ISSUE TITLE
Missed or badly triggered computed Actions: BUY and SELL. Significant Difference Between Backtest and LIVE.

### ISSUE THREAD

**Issue by Axel-CH** [freqtrade/freqtrade#746]
**## Step 2: Describe your environment**

  * Python Version: Python 3.6.4
  * Branch: Develop
  * Last Commit ID: b0536db (`git log --format="%H" -n 1`)
 
**## Step 3: Describe the problem:**
Hi everyone, I decided to finaly create an issue because i'm struggling with a problem since a few weeks:
There is an important difference between the signal that are computed on Backtest and the reality of actions in Live.
After a few observation I can tell that this is not related to a normal lag due to the network. The behavior seems to occur when a trigger is slightly triggered. 
**Exemple:** 
Sell When "Close Under Open Of Previous Candle"
dataframe['close'] < dataframe['open'].shift()
_Theoricaly Failing case_
0.00005155 < 0.00005156
(Those values are too close, the condition is not fullfilled)
_Theoricaly Working case_
0.00005155 < 0.00005160
(the gap between the two value is large enought, the condition is fullfilled)

### Steps to reproduce:

  1. Use 1 minute Time frame
  2. launch freqtrade on live with a strategy that use directly price or averages to trigger BUY or SELL actions
  3. Wait for a price movement that "slightly" correspond to the signal (BUY or SELL)
  4. The signal is not triggered or is badly triggered .
  5. When a few moment later there is a bigger price move that trigger the signal more Frankly, the signal 
      is triggered.
  
### Observed Results:

  * What happened?
The signal is not triggered sometimes, when there is a small gap between compared values that are used to take the descision
  * What did you expect to happen?
The signal should be triggered in live more or less at the same moment that is computed on the backtest (less than 1min drift tolerated on live)

### Relevant code exceptions or logs:

<img width="1539" alt="orders-missed-issue" src="https://user-images.githubusercontent.com/7593404/40408477-a63a0164-5e35-11e8-9939-ce30141efae4.png">
<img width="1030" alt="orders-missed-issue-computed-on-backtest-poa" src="https://user-images.githubusercontent.com/7593404/40410503-feb14248-5e3b-11e8-9184-e6827215acbf.png">
Note: On plot-dataframe I modified the code to display only the SELLs that are consecutive to BUYs, to have more clarity.

### Suspected cause
Loss of prescision due to python floating point: the problem with that theory is that I don't understand why the backtest is working as expected. The problem should be also present in that case...

### What I tried to fix that issue

1- Convert all collumns related to triggers to Decimal
2- Rise Ta-lib prescision (https://github.com/mrjbq7/ta-lib/issues/151)

The issue is still present.

Thank you for your help.
Regards,
Axel

**Comment by xmatthias** [comment]
There is a similar issue here: #477

With Bittrex - there is a known issue with 1m tickers (it can have a delay of up to 3 minutes) - which is their API, not the fault of the bot. Backtesting has the data always available - which will lead to a more reliable result.

I don't think that the "loss of precision" is the root cause for this problem - if it would be, backtesting would suffer from the same exact problem - as the floating error would be introduced there, too.

**Comment by Axel-CH** [comment]
Hi xmatthias,
Thank you for your answer, I'm agree with you about the floationg point.
I did'nt studied exactly how the Live data is handled by freqtrade. But the source of this issue could be a problem of sync between the data source API and freqtrade.
Explanation:
If there is a 10s difference between the "real" close and what freqtrade consider as close. Maybe It can cause that kind of behavior. It could be seen as a drift between candles.
What do you think about that?

**Comment by vertti** [comment]
Great to get more eyes on this problem. It's a problem we have known exists for a long time: big difference between live ops and backtesting. Calculations used in backtesting are the same as in live ops. The only difference I know of and suspect as at least one issue is that backtesting only has close prices and will use those to buy and sell. But live op calls a diffent api to get the current price to make bid. If those differ considerably then but will make very different decisions.

**Comment by FrankBloo** [comment]
Just as you guys, i'm also struggling with the signficant difference. 

I'm far from a python expert, just a beginner,  but I noticed that the Analyze.py --> parse_ticker_dataframe(ticker: list) function uses the columns  ['date', 'open', 'high', 'low', 'close', 'volume']. On the other hand, the backtesting.py --> backtest(self, args: Dict) uses the columns/headers = ['date', 'buy', 'open', 'close', 'sell']. I.e. it lacks the high and low. 

Is this actually true or has the backtest scrypt actually access to the high and low values?

**Comment by FrankBloo** [comment]
It seems that the backtest doesn't take into account when the minimum roi has been reached.

**Comment by creslinux** [comment]
I've not looked into the code here, so apologies for the open question. 
Are we sure back testing is not repainting? Its the most common cause of strategies performing better than in live.  
If I know how a 15minute candle will close on entering the period in backtest ill always make profit.
Backtest *MUST* only know about the current 15min candle *AFTER* it has closed  - it must be blind to tick data within that candl  to be accurate. The same applies on any timescale.

**Comment by vertti** [comment]
@FrankBloo where did you get that idea? It most certainly does consider that.

@creslinux feel free to check the backtesting loop. Would love to find a bug from there.

**Comment by xmatthias** [comment]
i think @creslinux has a good point here
we pass the whole candle in for the whole history.
in reality, the last candle will not be complete / move with each iteration - so the "close" price will move around until the candle is closed.

I would not know how to code this differently efficiently
one way would be as follows (assume we're using 5min ticker)
- filling the latest candle with 1m values (one after the other ...) for backtesting per ticker interval 
- backtest in a loop of time moving forward - each time removing all future data and filling the last candle with 1m data
- but that would have backtesting take forever as it would not evaluate the signals once - but once per minute (irrelevant of which ticker-interval is used).

it could improve backtesting vs. wet accuracy though

**Comment by jblestang** [comment]
If the backtest behaviour is ok, why not make the bot work like the backtest and not modify the bot to work like backtest?

**Comment by vertti** [comment]
Close price will be between the high and low, so we could simulate bot behaviour in several ways:
- run it by always selling with low or high instead of close
- run it by selling with random value between low or high
- run it three times by selling on high, then low, then close and umm.. average the results?

At least it would be good to add a parameter to simulate fluctuations, so default would be still selling with close price and with the flag enabled it would sell with random value between high and low. So you could test your strat as you normally would and then run it several times with the fluctuations enabled to test how it handles the fluctuations.

Or we could change when or how the bot picks the price. Like @jblestang says, we could time the bot to sell only if the price is very close to the previous close price. Or if it's higher than the previous close.

**Comment by creslinux** [comment]
If make a strategy decision within a bar period in backtest you will see a delta to live, unless backtesting full tickdata. 

The temptation is to see a candle from 13:00 to 13:15 Opens/Closes at $200/$100 and in backtest sell at 13:00 because the strategy says sell under $190.  
In live you'll not make that sell decision till 13:15:01, at which point the price is $100  - much lower than hoped.

**Comment by xmatthias** [comment]
why would it wait til 13:15 in live?
live does not wait for the full candle to decide - but uses "current rate" for the evaluation ...

**Comment by creslinux** [comment]
Depends on the strategy, MAs as example would use the close price

If you were to sell at a candle dipped under 190 it could overtrade 
as example  OHLC 200-250-100-220, you would not have wanted to exit there

The bot can't make a decision on a close price till after the candle has closed or testing will be off between 
If look at data from smaller timescales within a candle same rules apply, a 15 split into 3 5mins can only be seen at 5.01, 10.01 and along with the 15 at 15:01. This will break a lot of MAs based strategies to the fake profitable results unless very careful how implement

**Comment by xmatthias** [comment]
assuming your sample for OHLC 200-250-100-220 for the 15 minute period of 13:00 - 13:15.
in case of backtesting - the bot would never know when high / low happened - and therefore needs to either use a random value between ... or high or low or close.

In live/dry-mode however, it knows the current value at any given point in time - so it can (try) to sell at high.

Trying to simulate this is difficult as we don't have half-closed candle data - i guess @vertti 's randomness should provide a good mix?

Question that just jumped my mind - should backtesting use "low" to decide selling due to stop-loss? that's what the bot would probably do in real-time ...

**Comment by creslinux** [comment]
i think that will break all the MA strategies, you can't use close data until after close. 
How can we use some in-between random data, if we say its between open and close then we've already looked into the future which is broken.

**Comment by creslinux** [comment]
if want to use half closed data then the strategy should be using 7.5minute bars. 
That will then lead to the same discussion around introducing data at 3.25 minutes - its introducing repainting into backtest which should be avoided to prevent false positives.

We'll end up with buys and sells on MAs that would never use mid bar data and will plot cleanly missing any buy or sell range.  Those, and id argue all, can only be plotted using close of the previous time bar.

**Comment by jblestang** [comment]
The same apply for volume and rsi/adx on partial canddles

**Comment by creslinux** [comment]
I think what the project wants is to be able to paper test, quickly
Paper testing isn't back testing, to introduce repainting into backtest would break a lot of indicators unless exceptional care was applied to many different indicators.  The tradeoff of back testing is that you can only place your buy/sell AFTER the prior candle has closed. 

This is a decent link that described the challenges tradingview had with pinev2 and repainting. 
https://backtest-rookies.com/2017/11/29/tradingview-indicator-repainting/

I would proposed separate datasource and "paper-test" function be considered to allow fast paper testing.   As opposed to running in dry-run against live order book over days and weeks. 

Some data is available and more could be collected.   This link has CSVs with orderbook data for many years of BTC data, as example coinbaseUSD extract with 43million records. 
timestamp  / price / volume 
```
1417412423,300.000000000000,0.010000000000
1417415048,370.000000000000,0.010000000000
1417416612,370.000000000000,0.026555540000
1417498166,377.000000000000,0.010000000000
1417517949,377.750000000000,0.250000000000
1417517949,378.000000000000,3.750000000000
1417518257,378.000000000000,4.900000000000
1417518340,378.000000000000,5.200000000000
1417545780,378.000000000000,0.100000000000
 43883194
```

**Comment by Axel-CH** [comment]
I think that waiting for the candle to close to take a descision in live trading can be good or bad, depending to the situation. So I propose the folowing:

- Give the user the posibility to choose between the curent price as "close" or use the "real close"
- Add a "currentprice" field in the dataframe to alow the user to use that data directly as it is.

@vertti  Let me know what you think.

**Comment by arudov** [comment]
I have run into the same issue.
I have noticed, and I think this is the main issue why backtesting results are not equal to the bot real run results. The backtesting uses the information about closed candles, but the bot in live receives incomplete last candle. For example, the bot uses 15 mins tickers and it requests candles data once a minute, so it will receive 15 different informations about the last candle.

Here is the real example: the information about the last candle, Binance, ETHBTC, 15 min ticker interval, the requests were done with 1 min interval:
```
  [
    1528227900000, // times are the same
    "0.07980400",
    "0.07993300",
    "0.07966000",
    "0.07981000", // close price is different
    "389.24500000",
    1528228799999,
    "31.07320967",
    656,
    "243.22800000",
    "19.42391131",
    "0"
  ]
in 1 min:
    [
    1528227900000,
    "0.07980400",
    "0.07993300",
    "0.07966000",
    "0.07978200",
    "396.70500000",
    1528228799999,
    "31.66824611",
    705,
    "246.85900000",
    "19.71367394",
    "0"
  ]
in one more min
    [
    1528227900000,
    "0.07980400",
    "0.07993300",
    "0.07966000",
    "0.07979000",
    "779.99100000",
    1528228799999,
    "62.23541203",
    973,
    "536.22200000",
    "42.79462408",
    "0"
  ]
```
(you can check it with link https://api.binance.com/api/v1/klines?symbol=ETHBTC&interval=15m)

The only way we can fix it is to ignore the last candle during the real life bot execution. But it could affect bot results too, because buy or sell requests could not be satisfied because of the time delta between close price in the last candle and the current real price.

**Comment by xmatthias** [comment]
@arudov wouldn't this problem be fixed if the last candle from 15m ticker is ignored in backtesting, but instead filled up with 1m candles to get a "minutely" result?

ROI is based on "current rate" ... which would be closer to 1m close candles.

ignoring the latest candle in live trading would not fix the issue - the "current rate" would still be used to sell.

**Comment by vertti** [comment]
# Current situation

I think the reason is now clear why backtesting and live ops differ:

- backtesting only uses full candles, live op uses what it gets from exchange which almost always has a partial candle as the last item (let's say a 5min candle set where only 2minutes have gone by so far)
- backtesting uses last close as bid/ask price, where in reality it could be anything between high and low price of the last candle

So in live operation:
- both the buy and sell signals can fluctuate during an incomplete candle because the last candle data changes
 - stoploss can fire because price went all the way to "low" of the last candle but backtesting ignores that and checks stoploss against last close price

# Options to remedy the difference

## Changing how backtesting works

We can add a `randomize` flag which changes the last used candle's `close` to be something between the `high` and `low` so it resembles one of the potential partial candles. This could be used to check the robustness of your strategy.

## Changing how live operation works

0. We could remove the last candle from the received data. This would mean the buy/sell signals are less random because the do not include a partial candle in the end. But it would also mean the information is old. The bigger the used timeframe, the bigger the problem (let's say 1h candles or 4h candles and suddenly you have no information for the last 4 hours).
0 . We could make the bot buy and sell in synchronized times, so it would only trade once when the candle is complete, not during the partial candle.
0. We could add checks that we only buy if ask price is lower than the previous close that we used to calculate the buy signal. Same with sell but the opposite.
0. We can allow the bot to trade the way it does and do a better `--realistic-simulation` to backtesting to simulate all this randomness and partial candles.

**Comment by Axel-CH** [comment]
@vertti thank you for this summary.
**For how live operation works i would vote:**

- 2: Sync bot with api source

- Bonus: Add "current_price" for bearing to the loss of fresh data. As choise 2 create also the problem of choice 1.

**Comment by jblestang** [comment]
I vote for 1 or 2 but I think they are close or very close. Waiting for the bot to have a complete ticker is like waiting for a new incomplete ticker. So basically do no use the  buy/sell event from the last row (those are the ones that may vary due to incomplete ticker). Am I missing something with solution 2?

**Comment by vertti** [comment]
@jblestang yeah, i think the difference is that with option 2, bot would trade once per candle, and option 1 would keep the current trade frequency of multiple times per candle. Both options would ignore the last candle (partial).

**Comment by creslinux** [comment]
Good Morning. 
I don't agree with idea that if using hour candles and waiting for the close the data is old. 
This is trading a 1 hour time frame.
To use data within that timeframe is to base a strategy on multiple timeframes. 

Trading multiple timeframes can be a valid strategy.  
Its often used for early insight for where a larger timeframe *may be heading with the understanding there would be a very large number of false signals, which a strategy may apply a score like system too for predictive pivot points as example.

If to take a 1hour and 5 minute resolution for sake of example with 12 datapoints within 1 larger frame. As opposed to a random value within the hour frame - but to the same end.

If we're to look at any MA, RSI, Stoch (or any lagging indicator which are 99% available) on a 5 minute timeframe there will be a huge number of cross-overs/signals to sell that would be ignored on 1hour.  **There are a few indicator that can use this with validity, they are uncommonly applied and predictive for pivot points etc.

If the 5 minute interim values are set as psuedo values for the 1hour frame and interpreted as such in backtest this is false - this is repainting by definition . 
A backtest strategy will sell or buy early and be in positions it would never be in in live - possibly many many times flip-flopping within the that hour. 
Would each of the 12 points be tested against the last hour close or the last 5 minute close? one way we're trading a 5 minute resolution, the other repainting a 1 hour resolution - both will provide false 1hour strategy test results. 

If a strategy wishes to evaluating data on different resolution timeframes they can/should load the candles of that lower resolution explicitly into their strategy - either as a stand alone 5 minute, not 1hour strategy - or in a multi-timeframe/resolution strategy.

In Pine 3 viewing multiple resolutions is managed through the "security" operand. Described here: 
https://www.tradingview.com/wiki/Context_Switching,_The_%E2%80%98security%E2%80%99_Function
And in their release notes and with some better description here: 
https://www.tradingview.com/wiki/Pine_Script:_Release_Notes#Repainting_issue

An indicator tends to be lookahead, a strategy next to never.
What this means is if were to view an indicator plotted; its past is a static line to the last bar close. There is then a look-ahead portion, this will be wavering up and down maybe thousands of times in an hour influenced by orderbook price until locked in. 

This thread is toying with treated strategic back testing candle data as paper testing against orderbook. There are flaws in this that will break more than is addresses. 

If wish to head down the route of using  orderbook data in strategy this this must be presented as a separate datasource of orderbook to the strategy and not candle data, which would impact 99% of indicators - the two are different things.

**Comment by creslinux** [comment]
Here is a pandas tick-data to candle data page: 
https://www.quantinsti.com/blog/tick-tick-ohlc-data-pandas-tutorial/

As we don't have tick(orderbook) data for use in backtesting, to represent a paper test environment, could this be used in reverse 

What i mean to say is make use of 1minute candle data to build a fake, but realistic at 60points in an hour, orderbook. 

Presenting 1hr timeframe candles on close for strategy/all-lagging-indicators and an oderbook/tickdata will then represent trading on live / paper testing.  

It will allow fast paper testing as have the fake/built-orderbook/tickdata available.

**Comment by xmatthias** [comment]
i'm not sure if changing the live operation would fully fix the problem: 
`current_rate` is used for roi and stop-loss evaluation in live mode:
https://github.com/freqtrade/freqtrade/blob/fb49d706d04501479aff8bc3935d216c8c7743af/freqtrade/freqtradebot.py#L444

in my understanding, current rate is a ~20s (or so) actual data tick which is unavailable in backtesting.

Assuming 1h candles (dropping the last candle) and using 1m close data for `current_rate` (not filling up the 1h candles!!) would be the most correct approach in my opinion.

(just noticed @creslinux  last answer - it actually aligns to what i intend to say here).

**Comment by vertti** [comment]
I'm agreeing with @xmatthias . I think first steps now would be:

- keep using ask/bid prices as currently but ignore the last candle when calculating indicators used in the strategy

-apply the same to the backtesting which would guarantee it has no look-ahead

- add some `--fake-orderbook-data` switch if you want backtesting to randomize the ask/bid prices in the range of high/low using the last candle (which is not available to the strategy)

**Comment by creslinux** [comment]
I think @xmatthias  managed to say what took me a page in a sentence.  I agree with his post.

**Comment by jblestang** [comment]
@xmatthias not need to change the way ROI is being handled I think (except in backtesting)
@vertti would that mean using the open price of the next candle after a buy/sell event in the backtest? That should be close to what would happen in real life no?

**Comment by xmatthias** [comment]
adding one thing here:
i think backtesting does not show how many trades are "still open" and at what rate at the end of the backtesting period.

i think this would also be interresting ... (5 trade still open, each with -5% ...)

**Comment by arudov** [comment]
I come up with the question: why do we want to make backtesting results equal to the real trading results? It would be cool, but not necessary, I think.
A backtesting results should show the minimum possible earned money for that period. Let's say it should show the pessimistic scenario. It means that you can test the strategy and than be sure that the real bot trading will generate more money than during backtesting.

I think it could be done by using min candle price as the current price during `sell` signal calculation.

But not sure how to correct `buy` signal for this case.

**Comment by creslinux** [comment]
Trading profitably relies on risk/reward money management. 
Backtest needs to be accurate to best support RR decisions. 

Higher gain trades are typically the higher risk, hiding profit would leave low reward to high risk. 
Finding a relatively low gain low risk moving average based strategy for a bot to auto trade to is quite common. These may be lost entirely.

**Comment by creslinux** [comment]
Hi @axel-ch 
If your candle analysis for symbols being watched is taking longer than candle/bar then that would definitely be a problem, you're never going to finish.

There are a few elements in this.

At the moment everything works in a serial loop, download candle, analysis; and symbols with balance in a base currency are put to the front of the queue each loop.  Long story short it can be minutes before a pair at the back of your queue see any cpu analysis.

This has been addressed with multi proc/thread poc, and is being firmed up for Dev as an aiohttp/asyncio PR. 
(Reduces loop time from minutes to a few seconds)

A second problem is the loop processes the same candle data, hundreds to thousands of times,( if the symbol is one being put to the front of the queue) within a timeframe/candle

This is fixed in #1119 
Which stops expensive strat analysis being applied to the same dataframe continually within a candle.

**Comment by xmatthias** [comment]
I'll close this there was no activity for almost a year, and i think we've had a lot of progress in this regard, which was proven lately with  #1928 and #1931 (especially https://github.com/freqtrade/freqtrade/pull/1931#pullrequestreview-249874997) which matches dry-run and backtesting.
